src/main/java/org/zalando/nakadi/repository/MultiTimelineEventConsumer.java

```diff
@@ -0,0 +1,264 @@
+package org.zalando.nakadi.repository;
+
+import com.google.common.base.Preconditions;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.ListIterator;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Consumer;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.zalando.nakadi.domain.ConsumedEvent;
+import org.zalando.nakadi.domain.EventTypePartition;
+import org.zalando.nakadi.domain.NakadiCursor;
+import org.zalando.nakadi.domain.PartitionStatistics;
+import org.zalando.nakadi.domain.Timeline;
+import org.zalando.nakadi.domain.TopicPartition;
+import org.zalando.nakadi.exceptions.InvalidCursorException;
+import org.zalando.nakadi.exceptions.NakadiException;
+import org.zalando.nakadi.exceptions.NakadiRuntimeException;
+import org.zalando.nakadi.exceptions.ServiceUnavailableException;
+import org.zalando.nakadi.service.timeline.TimelineService;
+import org.zalando.nakadi.service.timeline.TimelineSync;
+import org.zalando.nakadi.util.NakadiCollectionUtils;
+
+public class MultiTimelineEventConsumer implements EventConsumer {
+    private final String clientId;
+    private final Map<EventTypePartition, NakadiCursor> latestOffsets = new HashMap<>();
+    private final Map<String, List<Timeline>> consumedEventTypes = new HashMap<>();
+    private final Map<String, TimelineSync.ListenerRegistration> timelineRefreshListeners = new HashMap<>();
+    private final Map<TopicRepository, EventConsumer> eventConsumers = new HashMap<>();
+    private final LinkedList<ConsumedEvent> eventsQueue = new LinkedList<>();
+    private final Map<EventTypePartition, String> borderOffsets = new HashMap<>();
```

Comment:
There are some new abstractions in this class which are not trivial. Could you clarify what `borderOffsets` is about? Does it hold the offsets pointing to the newest or oldest event in this partition?

If it's the latest available offset for that partition, I would suggest renaming to `latestAvailableOffset` to make it more clear or adding some comments. I find comments very useful when reading code, because I don't need to read the entire class to understand what is the meaning of something.